File: 2024-02-10-redis-publisher-and-subscriber-spring-boot/action-in-blog/publisher/src/main/java/blog/in/action/controller/TodoController.java
package blog.in.action.controller;

import blog.in.action.domain.Channel;
import blog.in.action.domain.EventLog;
import blog.in.action.publisher.EventLogPublisher;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

@RestController
@RequestMapping("/todos")
public class TodoController {

    private static final AtomicLong key = new AtomicLong(0);
    private static final Map<Long, String> data = new ConcurrentHashMap<>();

    static {
        data.put(key.incrementAndGet(), "Hello");
        data.put(key.incrementAndGet(), "Redis");
        data.put(key.incrementAndGet(), "World");
    }

    private final EventLogPublisher eventLogPublisher;

    public TodoController(EventLogPublisher eventLogPublisher) {
        this.eventLogPublisher = eventLogPublisher;
    }

    @GetMapping
    public List<Map.Entry<Long, String>> todos(HttpServletRequest request) {
        eventLogPublisher.publish(Channel.TODO_READ, new EventLog(request.getRemoteAddr()));
        return data.entrySet()
                .stream()
                .sorted(Map.Entry.comparingByKey())
                .toList();
    }

    @PostMapping
    public void todos(HttpServletRequest request, @RequestBody String requestBody) {
        eventLogPublisher.publish(Channel.TODO_INSERT, new EventLog(request.getRemoteAddr()));
        data.put(key.incrementAndGet(), requestBody);
    }

    @PutMapping("/{id}")
    public void todos(
            HttpServletRequest request,
            @PathVariable Long id,
            @RequestBody String requestBody
    ) {
        eventLogPublisher.publish(Channel.TODO_UPDATE, new EventLog(request.getRemoteAddr()));
        data.put(id, requestBody);
    }

    @DeleteMapping("/{id}")
    public void todos(HttpServletRequest request, @PathVariable Long id) {
        eventLogPublisher.publish(Channel.TODO_DELETE, new EventLog(request.getRemoteAddr()));
        data.remove(id);
    }
}

File: 2024-10-01-dynamodb-crud-example-with-spring-boot/action-in-blog/src/main/kotlin/action/in/blog/controller/TodoController.kt
package action.`in`.blog.controller

import action.`in`.blog.controller.request.TodoRequest
import action.`in`.blog.controller.response.TodoResponse
import action.`in`.blog.repository.TodoRepository
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/todos")
class TodoController(val repository: TodoRepository) {

    @GetMapping
    fun getTodos(): List<TodoResponse> {
        return repository.getTodos()
            .map { TodoResponse.of(it) }
    }

    @GetMapping("/{id}")
    fun getTodo(@PathVariable id: String): TodoResponse {
        return TodoResponse.of(
            repository.getTodo(id)
        )
    }

    @PostMapping
    fun createTodo(@RequestBody requestBody: TodoRequest) {
        repository.createTodo(requestBody.toEntity())
    }

    @PutMapping("/{id}")
    fun updateTodo(
        @PathVariable id: String,
        @RequestBody requestBody: TodoRequest
    ) {
        repository.updateTodo(requestBody.toEntity(id))
    }

    @DeleteMapping("/{id}")
    fun deleteTodo(@PathVariable id: String) {
        repository.deleteTodo(id)
    }
}